### PR TITLE
[DNM ]mds : make cdir as item of scrub to support distributed directory scr…

### PR DESCRIPTION
**Pull request by simon-rock**
…ubing

Signed-off-by: Simon Gao <simon29rock@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/mds/MDCache.cc [hunk 1]
```diff
@@ -13058,7 +13065,27 @@ void MDCache::repair_inode_stats_work(MDRequestRef& mdr)
...
+    return;
+  }
+  
+  MutationImpl::LockOpVec lov;
+  lov.add_rdlock(&diri->dirfragtreelock);
+  lov.add_rdlock(&diri->nestlock);
+  lov.add_rdlock(&diri->filelock);
+  if (!mds->locker->acquire_locks(mdr, lov))
+    return;
+  mds->server->respond_to_request(mdr, 0);
```
Comment (ukernel) [inline src/mds/MDCache.cc:13087]: this does not work as expected. we should acquire locks, execute ValidationContinuation::_dirfrag(), then drop locks.

File: src/mds/ScrubStack.cc [hunk 2]
```diff
@@ -149,132 +165,118 @@ void ScrubStack::kick_off_scrubs()
...
-	  break;
-	}
+    std::list<frag_t> frags;
+    in->dirfragtree.get_leaves(frags);
+    dout(20) << __func__ << "recursive mode frags num :" << frags.size() << dendl;
+    for (std::list<frag_t>::iterator i = frags.begin();
+        i != frags.end();
+        ++i) {
+      if (in->scrub_info()->scrubing_dirfrag.count(*i)){
+	dout(20) << __func__ << "has scrubed :" << *i << dendl;
```
Comment (ukernel) [inline src/mds/ScrubStack.cc:245]: we never skip a frag in the loop. how this can happen?

File: src/mds/ScrubStack.cc [hunk 3]
```diff
@@ -49,51 +49,67 @@ std::ostream &operator<<(std::ostream &os, const ScrubStack::State &state) {
...
+  if ((in = dynamic_cast<CInode*>(it)))  {
+    dout(10) << __func__ << " with {" << *in << "}"
+            << ", on_finish=" << on_finish << ", top=" << top << dendl;
+    in->scrub_initialize(parent, header, on_finish);
+  }
+  else if ((dir = dynamic_cast<CDir*>(it))){
+    dout(10) << __func__ << " with {" << *dir << "}"
+            << ", on_finish=" << on_finish << ", top=" << top << dendl;
+    // The edge directory must be in memory
+    dir->scrub_initialize_data(header, on_finish);
```
Comment (ukernel) [inline src/mds/ScrubStack.cc:98]: why not just calling dir->scrub_initialize()

File: src/mds/ScrubStack.cc [hunk 4]
```diff
@@ -149,132 +165,118 @@ void ScrubStack::kick_off_scrubs()
...
-    assert (!*added_children); // can't do this if children are still pending
 
     // OK, so now I can... fire off a validate on the dir inode, and
     // when it completes, come through here again, noticing that we've
     // set a flag to indicate the validate happened, and 
     scrub_dir_inode_final(in);
   }
 
-  *terminal = all_frags_terminal;
   *done = all_frags_done;
```
Comment (ukernel) [inline src/mds/ScrubStack.cc:282]: with this change, *done is always true

**Comment by ukernel** [comment]
new PR https://github.com/ceph/ceph/pull/35749

**Comment by simon-rock** [comment]
All commits have been moved to a new PR #35749. so close it.

**Pull request closed without merging**